indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceInputFormat.java

```diff
@@ -116,4 +129,72 @@ public int compare(WindowedDataSegment s1, WindowedDataSegment s2)
   {
     return new DatasourceRecordReader();
   }
+
+  private final JobConf dummy = new JobConf();
+  private final FileInputFormat fio = new FileInputFormat()
+  {
+    @Override
+    protected boolean isSplitable(FileSystem fs, Path filename)
+    {
+      return false;
+    }
+
+    @Override
+    public org.apache.hadoop.mapred.RecordReader getRecordReader(
+        org.apache.hadoop.mapred.InputSplit inputSplit, JobConf jobConf, Reporter reporter
+    ) throws IOException
+    {
+      throw new UnsupportedOperationException();
+    }
+  };
+
+  private DatasourceInputSplit toDataSourceSplit(List<WindowedDataSegment> segments)
+  {
+    String[] locations = null;
+    try {
+      locations = getFrequentLocations(segments);
+    }
+    catch (Exception e) {
+      // ignore
+    }
+    return new DatasourceInputSplit(segments, locations);
+  }
+
+  private String[] getFrequentLocations(List<WindowedDataSegment> segments) throws IOException
+  {
+    final CountingMap<String> counter = new CountingMap<>();
+    for (WindowedDataSegment segment : segments) {
+      FileInputFormat.setInputPaths(dummy, new Path(JobHelper.getURIFromSegment(segment.getSegment())));
+      for (org.apache.hadoop.mapred.InputSplit split : fio.getSplits(dummy, 1)) {
+        for (String location : split.getLocations()) {
+          counter.add(location, 1);
+        }
+      }
+    }
+
+    final TreeSet<Pair<Long, String>> sorted = Sets.<Pair<Long, String>>newTreeSet(
+        new Comparator<Pair<Long, String>>()
+        {
+          @Override
+          public int compare(Pair<Long, String> o1, Pair<Long, String> o2)
+          {
+            int compare = o1.lhs.compareTo(o2.lhs);
+            if (compare != 0) {
+              compare = o1.rhs.compareTo(o2.rhs);
+            }
+            return -compare;
+          }
```

Comment:
It's same with `TreeMap` issue above. because treeset/map does not keep duplicated entry in it, it will remove hosts with same count.